### ISSUE TITLE
RFE: Support for DESCRIBE tuner status in SAT>IP client mode.

### ISSUE THREAD

**Issue by lars18th** [catalinii/minisatip#222]
Hi,

The SAT>IP client mode is great. In the current code the standard way for update the tuner status is implemented . You can see the function "satipc_rtcp_reply(...)" in "satipc.c", as it implements the code to parse "signal" data in the "tuner" section.

However, not all SAT>IP servers implements this way for sharing the tuner status. If you read the Standard Specification, you can view that two different modes exist:

- Section 3.5.15.1. Description (of RTCP Announcements): Using RTCP messages received over the RTP UDP port+1.
- Section 3.5.7 Listing available media streams (DESCRIBE) page 39: Using the parameter "a=" in the RTSP response to a DESCRIBE request.

The fist way is the PUSH mode (the server sends the status time to time), and the second is the POLLING mode (the client sends a command to the server for request the status).

The I suggest to support both modes, as not all servers (for example ALi implementations) sends the push data of the tuner status. My suggestion, for a simple implementation with small changes is this:

a) When sending the keep-alive message (OPTIONS) over the RTSP channel, send after it one tuner-status request (DESCRIBE) over the same channel. With this behaviour you periodically get the status of the tuner with minimal overhead.

b) When receive the response of any DESCRIBE command, parse the response for get the tuner status. Then update the status of the adapter.

You agree to implement this?

**Comment by lars18th** [comment]
Hi,

Here more info for implementing the get status in the POOLING mode:

- Reproducing a program with a SAT>IP client I open a raw TCP connection to my hardware SAT>IP server and I send this command (at any time);

```
DESCRIBE rtsp://192.168.1.80:554/ RTSP/1.0
CSeq: 1
Accept: application/sdp

```

Then I receive:

```
RTSP/1.0 200 OK
CSeq: 1
Date: Fri, May 30 2014 13:46:36 GMT
Content-Base: rtsp://192.168.1.80/
Content-Type: application/sdp
Content-Length: 241

v=0
o=- 1401206822267000 1 IN IP4 192.168.1.80
s=SatIPServer:1 1
t=0 0
m=video 0 RTP/AVP 33
c=IN IP4 0.0.0.0
a=control:stream=33
a=fmtp:33 ver=1.0;src=1;tuner=1,240,1,15,10744,h,dvbs,,,,22000,56;pids=0,400,401,402,404
a=sendonly
```

As you can view, the line "a=fmtp:33 ..." contains the status of the tuner. As the standard indicates:

```
a=fmtp:33 ver=<major>.<minor>;src=<srcID>;tuner=<feID>,<level>,<lock>,<quality>,<frequency>,<polarisation>,<system>,<type>,<pilots>,<roll_off>,<symbol_rate>,<fec_inner>;pids=<pid0>,…,<pidn>
```

- Signal level level Numerical value between 0 and 255
 An incoming L-band satellite signal of
 -25dBm corresponds to 224
 -65dBm corresponds to 32
 No signal corresponds to 0
 In the example: 240

- Signal quality quality Numerical value between 0 and 15
 Lowest value corresponds to highest error rate
 The value 15 shall correspond to
 -a BER lower than 2x10-4 after Viterbi for DVB-S
 -a PER lower than 10-7 for DVB-S2
 In the example: 15

My suggestion is sending one DESCRIBE request after each PLAY and keep-alive (OPTIONS) command. This garanties that you receive regulary the tuner status.

You agree?

**Comment by catalinii** [comment]
Yeah i think it sounds good, your satip server does not send the signal over rtpc ?

**Comment by lars18th** [comment]
Hi @catalinii ,

> your satip server does not send the signal over rtpc ?

Yes, it doesn't send over RTPC. ALi implementation is very scarse. :-(

However, I feel it can support the DESCRIBE mode. If you like, please, add this other method for getting tuner status data; but with low priority... as I'm not sure if the ALi server sends the correct data in the response to the DESCRIBE request (the data is there, but I'm not sure if it's real data!).

Regards!

**Comment by lars18th** [comment]
Hi @catalinii ,

Thank you! This commit really implements this requested functionality:
http://github.com/catalinii/minisatip/commit/484d8a78d09f62e47f9f624bcbf226bdaef29031

However, I have two comments/questions:

1) I see that the tuner status is only updated after the first keep-alive. So, why not ask the tuner (send DESCRIBE) when you start to send the PLAY command? Also, this indicates to minisatip if the frquency is locked or not ("tuner=x,xxx,1,xx,..." and you can act accordingly).

2) A question: When you get the tuner status with this mode from the SAT>IP server, you send the current tuner status to the SAT>IP client using the RTP info?

Regards!

**Comment by catalinii** [comment]
1) I was thinking about the same, that you should get the status right away.

2) The rtcp signal status is sent according to the values from ad->strength & co...

**Comment by lars18th** [comment]
Hi @catalinii ,

>1) I was thinking about the same, that you should get the status right away.
>
>2) The rtcp signal status is sent according to the values from ad->strength & co...

Then, if you do 2) it's required to implement 1), as only after the fist keep-alive you have info about the status of the tuner (otherwise, it's invalid or zero!). Please, update the status when tunning. ;)

**Comment by catalinii** [comment]
I will link this to #243 ... the solution seems the same...
